File: src/my_robot_commander_cpp/src/robot_commander.cpp
# include <rclcpp/rclcpp.hpp>
# include <moveit/move_group_interface/move_group_interface.hpp>
# include <example_interfaces/msg/bool.hpp>
# include <example_interfaces/msg/float64_multi_array.hpp>
# include <my_robot_interfaces/msg/pose_cmd.hpp>

using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;
using Bool = example_interfaces::msg::Bool;
using Float64MultiArray = example_interfaces::msg::Float64MultiArray; 
using PoseCmd = my_robot_interfaces::msg::PoseCmd;
using namespace std::placeholders;


class RobotCommander
{
public:
  RobotCommander(std::shared_ptr<rclcpp::Node> node)
  {
    node_ = node;
    arm_group_ = std::make_shared<MoveGroupInterface>(node_, "arm");
    arm_group_->setMaxVelocityScalingFactor(1.0);
    arm_group_->setMaxAccelerationScalingFactor(1.0);

    gripper_group_ = std::make_shared<MoveGroupInterface>(node_, "gripper");

    open_gripper_sub_ = node_->create_subscription<Bool>(
      "open_gripper", 10, std::bind(&RobotCommander::OpenGripperCallback, this, _1));

    joint_cmd_sub_ = node_->create_subscription<Float64MultiArray>(
      "joint_cmd", 10, std::bind(&RobotCommander::JointCmdCallback, this, _1));

    pose_cmd_sub_ = node_->create_subscription<PoseCmd>(
      "pose_cmd", 10, std::bind(&RobotCommander::PoseCmdCallback, this, _1));


  }

  void goToNamedTarget(const std::string &target_name)
  {
    arm_group_->setStartStateToCurrentState();
    arm_group_->setNamedTarget(target_name);
    planAndExecute(arm_group_);
  }

  void goToJointTarget(const std::vector<double> &joint_group_positions)
  {
    arm_group_->setStartStateToCurrentState();
    arm_group_->setJointValueTarget(joint_group_positions);
    planAndExecute(arm_group_);
  }

  void goToPoseTarget(double x, double y, double z, double roll, double pitch, double yaw, bool cartesian_path = false){
    tf2::Quaternion q;
    q.setRPY(roll, pitch, yaw);
    q = q.normalize();

    geometry_msgs::msg::PoseStamped target_pose;
    target_pose.header.frame_id = "base_link";
    target_pose.pose.position.x = x;
    target_pose.pose.position.y = y;
    target_pose.pose.position.z = z;
    target_pose.pose.orientation.x = q.getX();
    target_pose.pose.orientation.y = q.getY();
    target_pose.pose.orientation.z = q.getZ();
    target_pose.pose.orientation.w = q.getW();

    arm_group_->setStartStateToCurrentState();

    if (!cartesian_path){
      arm_group_->setPoseTarget(target_pose);
      planAndExecute(arm_group_);
    }
    else{
      std::vector<geometry_msgs::msg::Pose> waypoints;
      waypoints.push_back(target_pose.pose);
      moveit_msgs::msg::RobotTrajectory trajectory;
      double fraction = arm_group_->computeCartesianPath(waypoints, 0.01, trajectory);
      RCLCPP_INFO(node_->get_logger(), "Cartesian path (%.2f%% achieved)", fraction * 100.0);
      if (fraction == 1.0)
      {
        arm_group_->execute(trajectory);
        RCLCPP_INFO(node_->get_logger(), "Path computed successfully. Moving the arm.");
      }
      else
      {
        RCLCPP_WARN(node_->get_logger(), "Only able to compute %.0f%% of the path. Not executing.", fraction * 100.0);
      }
    }
  }

  void openGripper()
  {
    gripper_group_->setStartStateToCurrentState();
    gripper_group_->setNamedTarget("gripper_open");
    planAndExecute(gripper_group_);
  }
  void closeGripper()
  {
    gripper_group_->setStartStateToCurrentState();
    gripper_group_->setNamedTarget("gripper_close");
    planAndExecute(gripper_group_);
  }

  
private:

  void planAndExecute(const std::shared_ptr<MoveGroupInterface> &move_group)
  {
    MoveGroupInterface::Plan plan;
    bool success = (move_group->plan(plan) == moveit::core::MoveItErrorCode::SUCCESS);
    if (success)
    {
      move_group->execute(plan);
    }
  }

  void OpenGripperCallback(const Bool &msg)
  {
    if (msg.data) {
      openGripper();
    } else {
      closeGripper();
    }
  }

  void JointCmdCallback(const Float64MultiArray &msg)
  {
    auto joint_target = msg.data;
    
    if (joint_target.size() != arm_group_->getJointNames().size()) {
      RCLCPP_ERROR(node_->get_logger(), "Joint command size does not match arm group size.");
      return;
    }
    goToJointTarget(joint_target);
  }

  void PoseCmdCallback(const PoseCmd &msg)
  {
    goToPoseTarget(msg.x, msg.y, msg.z, msg.roll, msg.pitch, msg.yaw, msg.cartesian_path);
  }

  std::shared_ptr<rclcpp::Node> node_;
  std::shared_ptr<MoveGroupInterface> arm_group_;
  std::shared_ptr<MoveGroupInterface> gripper_group_;

  rclcpp::Subscription<Bool>::SharedPtr open_gripper_sub_;
  rclcpp::Subscription<Float64MultiArray>::SharedPtr joint_cmd_sub_;
  rclcpp::Subscription<PoseCmd>::SharedPtr pose_cmd_sub_;
  
}; 






int main(int argc, char **argv)
{
  rclcpp::init(argc, argv); // Initialize ROS 2 context
  auto node = std::make_shared<rclcpp::Node>("robot_commander");
  auto robot_commander = RobotCommander(node);
  rclcpp::spin(node); // Spin the node to process callbacks
  rclcpp::shutdown(); // Shutdown ROS 2 context
  return 0;

}

File: src/my_robot_commander_py/my_robot_commander_py/robot_commander.py
import rclpy
from rclpy.node import Node
from moveit.planning import MoveItPy, PlanningComponent
from moveit_configs_utils import MoveItConfigsBuilder
from moveit.core.robot_state import RobotState
from geometry_msgs.msg import PoseStamped
import tf_transformations

from example_interfaces.msg import Bool
from example_interfaces.msg import Float64MultiArray
from my_robot_interfaces.msg import PoseCmd

ROBOT_CONFIG = MoveItConfigsBuilder(robot_name="my_robot", package_name="my_robot_moveit_config")\
                                    .robot_description_semantic("config/my_robot.srdf", {"name": "my_robot"})\
                                    .to_dict()

ROBOT_CONFIG = { 
    **ROBOT_CONFIG,
    "planning_scene_monitor": {
            "name": "planning_scene_monitor",
            "robot_description": "robot_description",
            "joint_state_topic": "/joint_states",
            "attached_collision_object_topic": "/moveit_cpp/planning_scene_monitor",
            "publish_planning_scene_topic": "/moveit_cpp/publish_planning_scene",
            "monitored_planning_scene_topic": "/moveit_cpp/monitored_planning_scene",
            "wait_for_initial_state_timeout": 10.0,
        },
        "planning_pipelines": {
            "pipeline_names": ["ompl"]
        },
        "plan_request_params": {
            "planning_attempts": 1,
            "planning_pipeline": "ompl",
            "max_velocity_scaling_factor": 1.0,
            "max_acceleration_scaling_factor": 1.0
        },
        "ompl": {
            "planning_plugins": ["ompl_interface/OMPLPlanner"],
            "request_adapters": ["default_planning_request_adapters/ResolveConstraintFrames",
                            "default_planning_request_adapters/ValidateWorkspaceBounds",
                            "default_planning_request_adapters/CheckStartStateBounds",
                            "default_planning_request_adapters/CheckStartStateCollision"],
            "response_adapters": ["default_planning_response_adapters/AddTimeOptimalParameterization",
                             "default_planning_response_adapters/ValidateSolution",
                             "default_planning_response_adapters/DisplayMotionPath"],
            "start_state_max_bounds_error": 0.1
        }
}

class RobotCommander(Node):
    def __init__(self):
        super().__init__("robot_commander_node")
        self.robot_ = MoveItPy(
            node_name="robot_commander_py",
            config_dict=ROBOT_CONFIG
        )
        self.arm_ : PlanningComponent = self.robot_.get_planning_component("arm")
        self.gripper_ : PlanningComponent = self.robot_.get_planning_component("gripper")

        self.open_gripper_sub = self.create_subscription(
            msg_type=Bool,
            topic="open_gripper_py",
            callback=self.open_gripper_callback,
            qos_profile=10
        )

        self.joint_cmd_sub = self.create_subscription(
            msg_type=Float64MultiArray,
            topic="joint_cmd_py",
            callback=self.joint_cmd_callback,
            qos_profile=10
        )

        self.pose_cmd_sub = self.create_subscription(
            msg_type=PoseCmd,
            topic="pose_cmd_py",
            callback=self.pose_cmd_callback,
            qos_profile=10
        )

    def go_to_named_target(self, target_name):
        self.arm_.set_start_state_to_current_state()
        self.arm_.set_goal_state(configuration_name=target_name)
        self.plan_and_execute(self.arm_)

    def go_to_joint_target(self, joint_positions):   
        if len(joint_positions) != 6:
            self.get_logger().error("go_to_joint_target: joint_positions must be 6")
            return
        robot_state = RobotState(self.robot_.get_robot_model())     
        joint_values = {
            "joint1": joint_positions[0],
            "joint2": joint_positions[1],
            "joint3": joint_positions[2],
            "joint4": joint_positions[3],
            "joint5": joint_positions[4],
            "joint6": joint_positions[5],
        }
        robot_state.joint_positions = joint_values
        self.arm_.set_start_state_to_current_state()
        self.arm_.set_goal_state(robot_state=robot_state)
        self.plan_and_execute(self.arm_)

    def go_to_pose_target(self, x, y, z, roll, pitch, yaw):
        q_x, q_y, q_z, q_w = tf_transformations.quaternion_from_euler(roll, pitch, yaw)
        pose = PoseStamped()
        pose.header.frame_id = "base_link"
        pose.pose.position.x = x
        pose.pose.position.y = y
        pose.pose.position.z = z
        pose.pose.orientation.x = q_x
        pose.pose.orientation.y = q_y
        pose.pose.orientation.z = q_z
        pose.pose.orientation.w = q_w
        self.arm_.set_start_state_to_current_state()
        self.arm_.set_goal_state(pose_stamped_msg=pose, pose_link="tool_link")
        self.plan_and_execute(self.arm_)

    def open_gripper(self):
        self.gripper_.set_start_state_to_current_state()
        self.gripper_.set_goal_state(configuration_name="gripper_open")
        self.plan_and_execute(self.gripper_)

    def close_gripper(self):
        self.gripper_.set_start_state_to_current_state()
        self.gripper_.set_goal_state(configuration_name="gripper_close")
        self.plan_and_execute(self.gripper_)
        
    def plan_and_execute(self, interface):
        plan_result = interface.plan()
        if plan_result:
            self.get_logger().info("Plan success, and now executing trajectory")
            self.robot_.execute(plan_result.trajectory,controllers=[])
        else:
            self.get_logger().error("Plan failed, check the target whether it is in the workspace or the target is not colliding with the obstacles")

    def open_gripper_callback(self, msg: Bool):
        if msg.data:
            self.open_gripper()
        else:
            self.close_gripper()

    def joint_cmd_callback(self, msg: Float64MultiArray):
        if msg.data.__len__() != 6:
            self.get_logger().error(f"joint_cmd_callback: joint_cmd must be 6")
            return
        self.go_to_joint_target(msg.data)

    def pose_cmd_callback(self, msg: PoseCmd):
        self.go_to_pose_target(msg.x, msg.y, msg.z, msg.roll, msg.pitch, msg.yaw)




def main(args=None):
    rclpy.init(args=args)
    node = RobotCommander()
    rclpy.spin(node)
    rclpy.shutdown()